### ISSUE TITLE
Check for arithmetic collisions

### ISSUE THREAD

**Issue by ChrisJefferson** [rurban/smhasher#51]
I have recently discovered some hash functions perform poorly on arithmetic operations on hashes,

For example on the FNV1a family, if I hash 32-bit integers, then if a+b=c+d, then f(a)+f(b)=f(c)+f(d). I think testing some basic arithmetic properties of hashes might be interesting. I'm not sure the best way of going about adding new tests, is there a guide (and are you interested in new types of tests?)

**Comment by rurban** [comment]
Hmm, never taught of that. Mostly we hash strings not ints, so they are pointers. But if you come up with some tests, we would be very pleased of course.

**Comment by cris691** [comment]
@ChrisJefferson I think that's a really cool idea. Have you taken a look at how some other tests are written (say in `KeysetTest.cpp` or `main.cpp`, which are usually pretty short functions, which means they're usually simple, tho perhaps shortness doesn't mean that for every test), and had a stab at writing your arithmetic tests?

**Comment by rurban** [comment]
This is similar to the reversible test, and would need symbolic solver support, like cmbc. I played around with it, but nothing to create useful tests so far. Related to security.

**Comment by thomasahle** [comment]
What's the point of a test like this?
Having a hash function that satisfies linear relations like what you wrote is actually a great benefit!
You can make some really cool algorithms using such functions. See e.g. 2.1 of "Subquadratic Algorithms for 3SUM" http://erikdemaine.org/papers/3SUM_WADS2005/paper.pdf

**Comment by rurban** [comment]
Cool properties are not so cool if you can trivially reverse them, or cancel them out as with the CRC family.

**Comment by ChrisJefferson** [comment]
In many case, an easy (and fairly good way) of hashing an ordered list is to hash each element, then add the hashes. This actually has provable good properties, as long as your hash function is "good", and in this case, satisfying linear relations is the very worst thing.

**Comment by thomasahle** [comment]
If you hash (x,y) as `h(x)+h'(y)` where `h` and `h'` are each 2-independent/universal and independent of each other, then the result is a 2-independent hash of `(x,y)`.
You don't need any "non-linearity" properties.
If `h` and `h'` are the same hash function, no properties can save you from `h(x)+h(x)=2h(x)`.

I guess this all depends on what properties you want from your hash function.
For crypto, I'm sure linearity isn't great.
For algorithms having a collision-resistant hash function, that's also linear, can be a great thing.

I like that about smhasher, one can get a quick overview of which functions satisfy which properties, and pick the one that suits their use-case. Just don't pretend like all of these "properties" are necessarily a bad thing to always be avoided.

**Comment by ChrisJefferson** [comment]
@thomasahle : I never said it was a bad thing to be avoided in all situations, just that it would be useful for my purposes to avoid it (and maybe others would want to know too).

Also, I don't actually know if FNV1a is actually a linear hash, it might just satisfy the property more frequently than expected (I haven't checked).

Just for information, my personal use for hashing very large sets of objects. I can quickly add and remove one thing from the set, by adding (or subtracting) it's hash from the total, which is very useful to me. For large sets, I can't (easily) have a separate hash for everything in the set, but as sets contain no duplicates h(x)+h(x) won't happen.

**Comment by thomasahle** [comment]
@ChrisJefferson You are right. That seems like a very reasonable requirement in your situation.

Just in case you are interested, allow me to entertain the following hash function:
For some large prime `p` and a random number `a < p`, let `h(x) = a^x mod p`. Then you are guaranteed that two sums colide `h(x_1)+h(x_2)+... = h(y_1)+h(y_2)+...` with at most probability `2^b/p`, where `b` is the number of bits in your keys. Basically because a degree `d` polynomial only has `d` roots.

Here `a^x (mod p)` (modular exponentiation, not xor) can be computed in time `O(x)`.
I don't know if that's fast enough for your use-case. Just wanted to mention it in case you care more about being safe than having a very fast hash function.

**Comment by oertl** [comment]
@thomasahle I guess you meant `a^x (mod p)` can be computed in time `O(log X)`.

 @ChrisJefferson Another option for hashing an ordered collection: Sort the element hashes and hash the sorted list of element hashes to get the final hash for the ordered collection. The quality of the hash should be better compared to just summing up the element hashes. Of course, sorting is an `O(N log N)` operation, but might be faster than the approach @thomasahle proposed, which has a time complexity of `O(N log X)` and `log X` is essentially given by the number of bits of the hash.

**Comment by ChrisJefferson** [comment]
The problem with sorting is you can't then operate incrementally -- adding and remove elements from the set as time progresses. Assuming you have a "good hash", adding seems to be more than reliable enough (just to say, this isn't security, I wouldn't be trying to "spin my own" for anything where hash collisions would create bugs / security holes).
